### ISSUE TITLE
Performance Problem.

### ISSUE THREAD

**Issue by zacksiri** [cabol/nebulex#27]
I tried switching to nebulex from my hand-rolled version using redis. There was quite a performance hit. The 3 bottom results were before nebulex, and everything on top is nebulex results. 

<img width="244" alt="screen shot 2561-10-03 at 07 09 08" src="https://user-images.githubusercontent.com/134488/46383598-591cd780-c6db-11e8-83af-c1f6e9be971f.png">

**Comment by cabol** [comment]
@zacksiri please give me more details about your environment and config, for instance, how many nodes, what cache topology are you rolling out (partitioned through the distributed adapter maybe?), what operations are you testing, how are you running the tests, in other words, if you can share some example code would be the best.

On other hand, a single docker container, 1 core could be limited, remember one of the most important features about Erlang/Elixir is taking advantage of multicore architectures, so single core and on top of that a docker container, it might be a very limited node.

Other thing thing to keep in mind, the distributed adapter uses "Distributed Erlang/Elixir" and that is implemented on top of TCP, so in terms of performance, it might be comparable with any other optimized protocol that uses TCP (like Redis or Memcached). But anyways, I wouldn't expect too much difference. So if you can repeat the test with better resources for Elixir (Nebulex) boxes as you mentioned, it would be great, and also gime as much details as possible.

Thanks, I stay tuned!

**Comment by cabol** [comment]
@zacksiri in that case it is weird, maybe as you mentioned there might be something in the environment, check this link: https://github.com/cabol/nebulex_examples/tree/master/nebulex_bench

There you will see the last bench I ran using the local adapter, you can run the bench tests again in your environment to compare (also there is a description about the environment I used)

Also, you can check in your env running the local bench for Nebulex: `mix nebulex.bench`; take a look tho the [benchmarks section](https://github.com/cabol/nebulex#benchmarks)

Let me know!

**Comment by zacksiri** [comment]
Yeah I’m going to run some benches, with Redis the service is on a different node, but moving to nebulex means all the load is in 1 place so all the calls to the cache might add up, for that reason add to the response time. I will do some further research.

**Comment by cabol** [comment]
As far as I understand, you are using the local adapter right? So you are running the tests on a single node, which make sense, you are trying to bench Nebulex locally (single-node), which is fine. The problem is, the node you are using is super small, what I'd do is the same test, but using bigger instances or nodes. For example, in my laptop (4 cores, 8 GB RAM, 256 SSD) I got this bench results (running `mix nebulex.bench`):

```
## LocalBench
benchmark name  iterations   average time
has_key?           1000000   1.43 µs/op
update_counter     1000000   1.46 µs/op
get!               1000000   1.50 µs/op
get                1000000   1.57 µs/op
add                1000000   1.69 µs/op
delete             1000000   1.70 µs/op
take               1000000   1.75 µs/op
replace            1000000   1.95 µs/op
set                1000000   2.36 µs/op
get_and_update      500000   3.59 µs/op
update              500000   3.62 µs/op
size                100000   14.67 µs/op
set_many            100000   16.09 µs/op
```

I'd try to run the same but in a docker container, but meanwhile you can try the same, no need to move Nebulex to a single node, you can keep it on each node, but try to use bigger nodes, at least similar to the one is running Redis.

**Comment by zacksiri** [comment]
I’m going to recreate the node using a container and limiting the cpu to 1 core. To compare to your results. Haven’t had the time today, I’ll let you know what I find.

**Comment by cabol** [comment]
That sounds good, stay tuned !!

**Comment by zacksiri** [comment]
Hey here are the results for 1 core inside the container. it is significantly slower than when I run it on my laptop. on my local laptop I'm getting similar results to yours.

```
## LocalBench
benchmark name  iterations   average time
has_key?           1000000   1.84 µs/op
update_counter     1000000   1.85 µs/op
take               1000000   1.86 µs/op
get                1000000   2.03 µs/op
get!               1000000   2.14 µs/op
delete             1000000   2.28 µs/op
replace            1000000   2.51 µs/op
add                1000000   2.52 µs/op
set                1000000   2.52 µs/op
get_and_update      500000   4.91 µs/op
update              500000   5.40 µs/op
size                100000   16.65 µs/op
transaction         100000   18.09 µs/op
set_many            100000   21.08 µs/op
all                  10000   129.36 µs/op
get_many              1000   2463.41 µs/op
```

2 Cores

```
## LocalBench
benchmark name  iterations   average time
update_counter     1000000   2.00 µs/op
has_key?           1000000   2.28 µs/op
take               1000000   2.32 µs/op
get                1000000   2.48 µs/op
get!               1000000   2.52 µs/op
delete             1000000   2.60 µs/op
add                1000000   2.82 µs/op
set                1000000   2.91 µs/op
replace            1000000   2.99 µs/op
update              500000   5.46 µs/op
get_and_update      500000   5.87 µs/op
transaction         100000   19.99 µs/op
size                100000   20.31 µs/op
set_many            100000   21.25 µs/op
all                  10000   187.35 µs/op
get_many              1000   2719.68 µs/op
```

4 Cores

```
## LocalBench
benchmark name  iterations   average time
update_counter     1000000   1.95 µs/op
take               1000000   2.33 µs/op
has_key?           1000000   2.38 µs/op
get                1000000   2.53 µs/op
delete             1000000   2.54 µs/op
get!               1000000   2.66 µs/op
add                1000000   2.87 µs/op
set                1000000   2.87 µs/op
replace             500000   2.96 µs/op
get_and_update      500000   5.65 µs/op
update              500000   6.00 µs/op
size                100000   20.43 µs/op
transaction         100000   20.75 µs/op
set_many            100000   22.01 µs/op
all                  10000   170.18 µs/op
get_many               500   3037.67 µs/op
```

**Comment by zacksiri** [comment]
I just tried on 2 different cloud providers, I'm getting very similar results between them. this one I posted is actually better.

**Comment by cabol** [comment]
@zacksiri awesome, that's good to know, I like this kind of issues a lot, because they help you to improve or discover bugs, or new issues, but I'm glad to hear you were able to fix your issue. Any further issue(s) don't hesitate to ping me o r open a new issue.

On other hand I'm also glad Nebulex is being useful for you (that's the goal) and I'd really appreciate if you cover it in you Elixir Foundation, that would be great!!

And, I'm working very hard to release 1.0.0 version ASAP, there are a lot of improvements and new features.

Thank you very much for your post and findings, it was very fruitful :)

**Comment by zacksiri** [comment]
So just to complete the issue with a screenshot. 

everything above 54e5fc6 is with nebulex everything below is redis. Ignore the first request that says 156ms that has to make the initial network call to cache a token, everyhing after that call is using nebulex cache.

![screen shot 2018-10-09 at 20 54 26](https://user-images.githubusercontent.com/134488/46674142-a16d5580-cc05-11e8-93a1-561a9e359567.png)

**Comment by cabol** [comment]
:thumbsup:
